README
Gocaml v1.0
-----------

Author: adb@tardis.ed.ac.uk (Andrew Birkett)
Date: 3 March 2002

This is gocaml, a graphical frontend for GMP (Go Modem Protocol)
engines.  Although usable, it has a very minimal user interface and is
intended more as an example of ocaml programming than a serious Go
client.  NOTE: It does not provide a computer player itself - it needs
something like gnugo for that.

It was written to explore the ocaml language, and also because I
couldn't get cgoban to run on my linux IPAQ.  It was almost completely
written natively on the IPAQ (with folding keyboard) using emacs and
ocamlc.  Great performance, great language.

See http://www.tardis.ed.ac.uk/~adb/gocaml/index.html for more info.

Installation
------------

First install the ocaml compiler from http://caml.inria.fr/ocaml.  Gocaml
was developed using ocaml-3.01, but any more recent version should be
fine.

Next, install gnugo from http://www.gnu.org/directory/gnugo.html and
make sure you can launch it fine by typing 'gnugo' at your shell
prompt.

Finally, build gocaml by typing 'make all'.  This will produce the 'gocaml'
executable.  Try './gocaml -help'.

Running Gocaml
--------------

By default, gocaml starts with a 9x9 board, no handicap and launching
gnugo with some sensible flags.  You are playing black, so click
somewhere to make your move.  Then you'll see gnugo's white move.  The
last stone to be played is marked with a dot.  Hitting any key passes.
The game ends when both sides pass.  

Currently, gocaml doesn't score the game so you have to look at the
'log-gnugo.txt' file to see what gnugo thought the score was.

By default, gnugo will write out a game file into log-sgf.txt.  Gocaml
will write it's own game record (non-SGF) into log-moves.txt.  A
record of the GMP communications goes into log-gmp.txt.

The following options are available:

  -handicap sets the handicap.  Valid arguments are 2 to 9.
  -color sets the human player color.  Valid arguments are 'black' or 'white'
  -play_local means both sides will be humans, and take turns about clicking.
  -board_size can be 9, 11, 13 or 19.
  -winsize sets the height and width of the GUI window.  Due to limitations in
    the ocaml graphics library, you can't usefully resize the window.
  -gmp_prog sets the command line used to invoke the GMP engine.  The default
    is gnugo --mode=gmp --quiet -o log-sgf.txt --score all 2>log-gnugo.txt

Design
------

The board (board.ml) is presented as an abstract type, but is implemented
as a map from board intersection to color.  All of the clique/liberty logic
is in the Board module.

We keep track of previous board states using the History module
(history.ml).  Board states are hashed to allow fast comparisons when
checking for Ko violations.

The Gui module has a straightforward interface (print board and ask
for move) but a basic implementation using the Graphics library.  It 
would be easy to add alternative GUI implementation.

The GMP implementation is provided in two layers.  Firstly, the Rawgmp
module provides marshalling/unmarshalling of GMP messages but doesn't
worry much about the semantics of the data.  Secondly, the Gmp module
adds sequence number checking and presents a higher level interface
(wait for ok, etc).

Bugs
----

Yes.

Todo
----

* Scoring

* Make GMP implementation more robust
  - detect start bit properly
  - handle talk text properly
  - active start (gocaml sends NewGame)
  - handle conflicts properly
  - handle DENY properly

* Better GUI feedback on mousedown

Copyrights and License
----------------------

All files are under the GNU General Public License, which may be found
in the file COPYING.



CHANGES
Gocaml v1.0
-----------

Initial public release:
  - Either local (human vs human) and remote (human vs GMP) modes
  - Play black or white
  - Any board size (9x9, 11x11, 13x13 or 19x19)
  - Handicaps (2 to 9)
  - Uses GMP (Go Modem Protocol) to communicate with game engines.
  - Simple GUI interface (designed for use on Compaq IPAQ)


main.ml
open Common
(*open Rawgmp *)

exception User_exit

type player_type = Local | Remote of string

let gmp_prog = "gnugo --mode=gmp --quiet -o log-sgf.txt --score all 2>log-gnugo.txt"
let handicap = ref None
let color : color ref = ref `Black
let white_player = ref (Remote gmp_prog)
let black_player = ref Local
let board_size = ref B9;;
let winsize = ref 700;;

let options = 
  [("-handicap", 
    Arg.Int (fun n -> 
      if n < 2 then raise (Arg.Bad "Handicap must be 2 or greater")
      else handicap := (Some n)),
    "Sets blacks advantage");

   ("-color",
    Arg.String (fun s -> match s with
      "white" -> color := `White
    | "black" -> color := `Black
    | _ -> raise (Arg.Bad "Color must be 'black' or 'white'")), 
    "Chooses local color");

   ("-local",
    Arg.Unit (fun () -> white_player := Local; black_player := Local),
    "Run gocaml without a remote player");

   ("-boardsize",
    Arg.Int (fun n -> match n with
      9 | 11 | 13 | 19 -> board_size := boardsize_of_int n
    | _ -> raise (Arg.Bad "Board size must be 9, 11, 13 or 19")),
    "Sets the board size");

   ("-winsize", Arg.Int (fun n -> winsize := n), 
    "Sets window dimensions");

   ("-gmpprog", Arg.String (fun s -> white_player := Remote s), 
    ("Sets GMP-compatible program to use (default is " ^ gmp_prog));
];;

let errmsg = "Gocaml v1.0.\n\nCommand line options are:";;

let parse_command_line () =
  Arg.parse options ignore errmsg

let rec rmt_move c =
  try 
    begin
      match Gmp.receive_move c with
	Move (p,_) -> History.make_turn p
      | Pass _ -> History.make_pass ();
    end;
    Gmp.send_ok c; 
  with 
    Gmp.UnexpectedCommand cmd ->
      Rawgmp.add_to_log ("Expecting MoveCmd, but got " ^ Rawgmp.command_to_string cmd ^ ".  Now waiting for move again");
      rmt_move c 
  | History.Ko ->
      Rawgmp.add_to_log "Denying last move due to Ko, now waiting for new remote move";
      Gmp.send_deny c;
      rmt_move c 

let rec lcl_move () =
  try
    match Gui.ask_for_move () with 
      Click p -> History.make_turn p
    | PassMove -> History.make_pass ()
    | Quit -> raise User_exit
  with
    Board.Illegal_move (n, p) -> Printf.printf "Illegal: %s \n" n; lcl_move ()
  | History.Ko -> print_string "Ko!\n"; lcl_move ()

let xmit_gmp c =
  (match !History.last_move with
    None -> ()
  | Some m -> Gmp.send_move c m);
  Gmp.wait_for_ok c true

let gameloop (wi, wo) (bi, bo) =
(*
  let bi () = Printf.printf "<BI>"; bi (); Printf.printf "</BI>\n"
  and bo () = Printf.printf "<BO>"; bo (); Printf.printf "</BO>\n"  
  and wi () = Printf.printf "<WI>"; wi (); Printf.printf "</WI>\n"
  and wo () = Printf.printf "<WO>"; wo (); Printf.printf "</WO>\n"
  in
*)
  try 
    let rec inner_loop () =
      if History.twopass () then History.end_info ()
      else begin
	Gui.draw_board ();
	match !History.next_turn with
	  `Black -> bi (); wo (); inner_loop ()
	| `White -> wi (); bo (); inner_loop ()
      end
    in
    inner_loop ()
  with 
    User_exit -> History.end_info ()

let passive_start gmp = 
  match Gmp.read_command gmp with
    Rawgmp.NewGameCmd ->
      Gmp.send_ok gmp;
      Gmp.respond_to_queries gmp;
  | m -> failwith ("Expecting NewGameCmd, but got " ^ Rawgmp.command_to_string m ^ " instead.")

let rec score_loop (b, captures) =
  Gui.draw_end_board b captures;
  match Gui.ask_for_move () with
  | PassMove | Quit -> ()
  | Click p -> 
      Board.forcetoggle b p;
      score_loop (b, captures)
	
let main () =
  parse_command_line ();
  let iofuncs = function 
      Local -> (lcl_move , (fun () -> ()))
    | Remote s -> 
	let gmp = Gmp.init_connection (Unix.open_process s) in
	passive_start gmp;
	((fun () -> rmt_move gmp), (fun () -> xmit_gmp gmp)) 
  in
  Gui.init_gui !winsize !board_size;
  History.init_history !handicap !board_size;
  Gui.draw_board ();
  gameloop (iofuncs !white_player) (iofuncs !black_player) |>
    score_loop
    
let _ = main ();;
